sql: add crdb_internal.validate_multi_region_zone_configs builtin

otan opened the pull request:
In addition to the release note, added some pgcodes to validation errors
as well as minor refactoring to make it work and error nicely with the
"overriding" error message.

Release note (sql change): Add a new builtin
`crdb_internal.validate_multi_region_zone_configs` which validates all
zone configurations in the current database is correctly configured.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/62780)
<!-- Reviewable:end -->

ajstorm reviewed (approved):
Nit: In release note - "which validates all zone configurations in the current database _are_ correctly configured"

> Reviewed 5 of 8 files at r1, 3 of 3 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62780#-:-MX2Nm4UCr3Xu4IvQ_PG:ba2xbnm)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy and @otan)

___
*[docs/generated/sql/functions.md, line 881 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/62780#-MX2U9iWCUfe5rzoAq7d:-MX2U9iWCUfe5rzoAq7e:b4xxdys) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e1b61d0bd449f130d1e34f332b53b3a07460ad7/docs/generated/sql/functions.md#L881)):*
> ```Markdown
> table’s zone configuration this will return NULL.</p>
> </span></td></tr>
> <tr><td><a name="crdb_internal.validate_multi_region_zone_configs"></a><code>crdb_internal.validate_multi_region_zone_configs() &rarr; <a href="bool.html">bool</a></code></td><td><span class="funcdesc"><p>Validates all multi-region zone configurations are correctly set
> ```

Nit: I believe this should be "setup", not "set up".
___
*[pkg/ccl/logictestccl/testdata/logic_test/multi_region_zone_configs, line 42 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/62780#-MX2hoIf8bZGwhVGTjsf:-MX2hoIf8bZGwhVGTjsg:b-9woie4) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e1b61d0bd449f130d1e34f332b53b3a07460ad7/pkg/ccl/logictestccl/testdata/logic_test/multi_region_zone_configs#L42)):*
> ```
> 
> statement ok
> SELECT crdb_internal.validate_multi_region_zone_configs()
> ```

FYI: Longer term, does this mean that we should be removing the calls to `SHOW ZONE CONFIGURATION` in these tests?  Or do you think that there's still value in having them so that the tests are more visually self-describing?
___
*[pkg/sql/region_util.go, line 834 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/62780#-MX2j0sjArmNbMrI0KkF:-MX2j0sjArmNbMrI0KkG:bp18i3m) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e1b61d0bd449f130d1e34f332b53b3a07460ad7/pkg/sql/region_util.go#L834)):*
> ```Go
> }
> 
> // ValidateCurrentDatabaseMultiRegionZoneConfigs is part of the tree.EvalDatabase interface.
> ```

Nit: maybe end this line with a comma?
___
*[pkg/sql/region_util.go, line 835 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/62780#-MX2jNaZ1TLTfc7AYcS1:-MX2jNaZ1TLTfc7AYcS2:bq7jl91) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e1b61d0bd449f130d1e34f332b53b3a07460ad7/pkg/sql/region_util.go#L835)):*
> ```Go
> 
> // ValidateCurrentDatabaseMultiRegionZoneConfigs is part of the tree.EvalDatabase interface.
> // and as such is intended for DML use. It returns an empty DatabaseRegionConfig
> ```

Nit: I think this should read "It returns nil..."
___
*[pkg/sql/region_util.go, line 837 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/62780#-MX2jfvx7NbLFDK4JLEO:-MX2jfvx7NbLFDK4JLEP:b3h9ell) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e1b61d0bd449f130d1e34f332b53b3a07460ad7/pkg/sql/region_util.go#L837)):*
> ```Go
> // and as such is intended for DML use. It returns an empty DatabaseRegionConfig
> // if the current database is not multi-region enabled.
> func (p *planner) ValidateCurrentDatabaseMultiRegionZoneConfigs(ctx context.Context) error {
> ```

Nit: Should there be a mention somewhere in the comment above that this also validates tables?  Alternatively, should the function name be changed to `ValidateMultRegionZoneConfigsForCurrentDatabaseAndTables`?
___
*[pkg/sql/region_util.go, line 842 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/62780#-MX2j83hBiHwCiqj5J2y:-MX2j83hBiHwCiqj5J2z:b-at74bi) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e1b61d0bd449f130d1e34f332b53b3a07460ad7/pkg/sql/region_util.go#L842)):*
> ```Go
> 		p.txn,
> 		p.CurrentDatabase(),
> 		tree.DatabaseLookupFlags{Required: true},
> ```

Nit: AvoidCached?  IncludeOffline?
___
*[pkg/sql/region_util.go, line 1172 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/62780#-MX2rsht4br4upM0LSSf:-MX2rsht4br4upM0LSSg:boprvr2) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e70529f84761b9b9b62a7f2cf0a07d45ee9c3a9/pkg/sql/region_util.go#L1172)):*
> ```Go
> 					// the ID.
> 					indexName := fmt.Sprintf("unknown with ID = %s",
> 						strconv.FormatUint(uint64(s.IndexID), 10))
> ```

I have to admit, I have not idea why I wrote it this way the first time :facepalm: 
___
*[pkg/sql/region_util.go, line 1185 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/62780#-MX2lI5X9pw3KA72lLZx:-MX2lI5X9pw3KA72lLZy:b-fma3ut) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e1b61d0bd449f130d1e34f332b53b3a07460ad7/pkg/sql/region_util.go#L1185)):*
> ```Go
> 
> // validateZoneConfigForMultiRegionDatabase validates that the zone config
> //for the databases matches the exact expectations as the multi-region
> ```

Nit: space at beginning of row.

Nit: I think we can probably remove "the  exact expectations as the", as it's somewhat implied.
___
*[pkg/sql/region_util.go, line 1212 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/62780#-MX2m7qeCbWMkyjSWORt:-MX2m7qeCbWMkyjSWORu:b-zbkq9) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e1b61d0bd449f130d1e34f332b53b3a07460ad7/pkg/sql/region_util.go#L1212)):*
> ```Go
> 	if !same {
> 		dbName := tree.Name(dbDesc.GetName())
> 		return validateZoneConfigForMultiRegionErrorHandler.newError(
> ```

Nice.  I like this encapsulation of the error message.
___
*[pkg/sql/region_util.go, line 1300 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/62780#-MX2svx82lCtjAuDcNzn:-MX2svx82lCtjAuDcNzo:br0xjqn) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e1b61d0bd449f130d1e34f332b53b3a07460ad7/pkg/sql/region_util.go#L1300)):*
> ```Go
> 					return validateZoneConfigForMultiRegionErrorHandler.newError(
> 						"index",
> 						fmt.Sprintf("%s@%s", tableName.String(), indexName),
> ```

Nit: Since we're using tableName.String() down here already, might want to use indexTreeName.String() so that we don't need `indexName` at all.  If you do this, might want to change indexTreeName to indexName.
___
*[pkg/sql/sem/builtins/builtins.go, line 5075 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/62780#-MX2w-XwEf3CE0973xji:-MX2w-XwEf3CE0973xjj:b4sk2pi) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e1b61d0bd449f130d1e34f332b53b3a07460ad7/pkg/sql/sem/builtins/builtins.go#L5075)):*
> ```Go
> 					evalCtx.Context,
> 				); err != nil {
> 					return nil, err
> ```

Should this return `false` here?


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62780)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm and @miretskiy)

___
*[pkg/ccl/logictestccl/testdata/logic_test/multi_region_zone_configs, line 42 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/62780#-MX2hoIf8bZGwhVGTjsf:-MX3muiY9Yw-ZLuBXcTr:b-fd5x5o) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e1b61d0bd449f130d1e34f332b53b3a07460ad7/pkg/ccl/logictestccl/testdata/logic_test/multi_region_zone_configs#L42)):*
<details><summary><i>Previously, ajstorm (Adam Storm) wrote…</i></summary><blockquote>

FYI: Longer term, does this mean that we should be removing the calls to `SHOW ZONE CONFIGURATION` in these tests?  Or do you think that there's still value in having them so that the tests are more visually self-describing?
</blockquote></details>

i reckon there's value in having both.
___
*[pkg/sql/region_util.go, line 1172 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/62780#-MX2rsht4br4upM0LSSf:-MX3qzsc8NL6BRgYDEbx:bb7hf0f) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e70529f84761b9b9b62a7f2cf0a07d45ee9c3a9/pkg/sql/region_util.go#L1172)):*
<details><summary><i>Previously, ajstorm (Adam Storm) wrote…</i></summary><blockquote>

I have to admit, I have not idea why I wrote it this way the first time :facepalm: 
</blockquote></details>

hehe
___
*[pkg/sql/region_util.go, line 1300 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/62780#-MX2svx82lCtjAuDcNzn:-MX3rK2t-5HoB6rHsO_q:b-a9mn2y) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e1b61d0bd449f130d1e34f332b53b3a07460ad7/pkg/sql/region_util.go#L1300)):*
<details><summary><i>Previously, ajstorm (Adam Storm) wrote…</i></summary><blockquote>

Nit: Since we're using tableName.String() down here already, might want to use indexTreeName.String() so that we don't need `indexName` at all.  If you do this, might want to change indexTreeName to indexName.
</blockquote></details>

i can't do `tree.Name(name).String()` in one pop because Name has a pointer receiver, so this has to be as it is.
___
*[pkg/sql/sem/builtins/builtins.go, line 5075 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/62780#-MX2w-XwEf3CE0973xji:-MX49kTS-Kliycr5yriN:bp6lba) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e1b61d0bd449f130d1e34f332b53b3a07460ad7/pkg/sql/sem/builtins/builtins.go#L5075)):*
<details><summary><i>Previously, ajstorm (Adam Storm) wrote…</i></summary><blockquote>

Should this return `false` here?
</blockquote></details>

the error will supercede all else :P 


<!-- Sent from Reviewable.io -->

arulajmani reviewed (changes requested):
For all this validation we seem to be either reading descriptors from the store or using the DDL variants when constructing zone configs -- is this what we want for a builtin? I think it should be fine to used leased descriptors for the builtin, right?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62780#-:-MX4e4SX0hNROb-tnwmV:b1hnoml)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm, @miretskiy, and @otan)

___
*[pkg/sql/region_util.go, line 842 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/62780#-MX2j83hBiHwCiqj5J2y:-MX4cdmPEf5Gni_fE6Oh:b-rt9pwo) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e1b61d0bd449f130d1e34f332b53b3a07460ad7/pkg/sql/region_util.go#L842)):*
<details><summary><i>Previously, ajstorm (Adam Storm) wrote…</i></summary><blockquote>

Nit: AvoidCached?  IncludeOffline?
</blockquote></details>

I don't fully understand, why do we need these flags here?
___
*[pkg/sql/region_util.go, line 888 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/62780#-MX4bbDW28ir5s1jMQQF:-MX4bbDW28ir5s1jMQQG:b-ozqps3) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff9533c86f6f8ac3530a3ecf9adcc42efddd5db4/pkg/sql/region_util.go#L888)):*
> ```Go
> 		tree.DatabaseLookupFlags{
> 			Required:       true,
> 			AvoidCached:    true,
> ```

Are you sure you want AvoidCached here? This method is called from the `DefaultToDatabasePrimaryRegionBuiltinName`, which means any insert into a Regional By Row table without an explicit value for `crdb_region`. A roundtrip here wouldn't fly, will it?


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62780)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm, @arulajmani, and @miretskiy)

___
*[pkg/sql/region_util.go, line 842 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/62780#-MX2j83hBiHwCiqj5J2y:-MX4mywQ66sq5Z1R10Et:ba0k5xh) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e1b61d0bd449f130d1e34f332b53b3a07460ad7/pkg/sql/region_util.go#L842)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

I don't fully understand, why do we need these flags here?
</blockquote></details>

Thinking about it more, yeah I don't think we do. We don't mind using the lease in the case, and no offline job code goes through this logic (yet).
___
*[pkg/sql/region_util.go, line 888 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/62780#-MX4bbDW28ir5s1jMQQF:-MX4lZ3C8QvADvb4UR-O:b-8371ok) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff9533c86f6f8ac3530a3ecf9adcc42efddd5db4/pkg/sql/region_util.go#L888)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Are you sure you want AvoidCached here? This method is called from the `DefaultToDatabasePrimaryRegionBuiltinName`, which means any insert into a Regional By Row table without an explicit value for `crdb_region`. A roundtrip here wouldn't fly, will it?
</blockquote></details>

oops, good catch, didn't intend this to be here.


<!-- Sent from Reviewable.io -->

otan committed:
sql: add crdb_internal.validate_multi_region_zone_configs builtin

In addition to the release note, added some pgcodes to validation errors
as well as minor refactoring to make it work and error nicely with the
"overriding" error message.

Also fix up some of the string escaping to use tree.Name.

Release note (sql change): Add a new builtin
`crdb_internal.validate_multi_region_zone_configs` which validates all
zone configurations in the current database are correctly configured.

arulajmani reviewed (approved):
I think @ajstorm already approved this, but <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> to me too after your changes 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62780#-:-MX54Bhu9jEhl8MOn49m:bulcjtr)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajstorm, @arulajmani, and @miretskiy)



<!-- Sent from Reviewable.io -->

otan commented:
bors r=ajstorm,arulajmani

otan commented:
bors r=ajstorm,arulajmani

ajstorm reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62780)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/sql/region_util.go, line 888 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/62780#-MX4bbDW28ir5s1jMQQF:-MX7AotV4KNb5YeWrJ13:beo5b8k) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff9533c86f6f8ac3530a3ecf9adcc42efddd5db4/pkg/sql/region_util.go#L888)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

oops, good catch, didn't intend this to be here.
</blockquote></details>

Good catch @arulajmani!


<!-- Sent from Reviewable.io -->
